pyccel/ast/headers.py

```diff
@@ -268,19 +269,11 @@ def template_resolve(arg_code, Tname, typ):
         #TODO: handle the case of functions arguments
 
         arg_codes = [list(arg_code) for arg_code in product(*dtypes)]
-        for Tname in templates:
-            i = 0
-            leng = len(arg_codes)
-            for i in range(leng):
-                for typ in templates[Tname].args:
-                    tmp_codes, resolved = template_resolve(arg_codes[0],
-                            Tname, typ)
-                    if resolved:
-                        arg_codes.append(tmp_codes)
-                if len(arg_codes) > leng:
-                    arg_codes.pop(0)
-                else:
-                    arg_codes.append(arg_codes.pop(0))
+        for tmplt in templates:
+            arg_codes = [process_template(arg_code, tmplt, d_type, j) for arg_code in arg_codes\
+                    for j, d_type, in enumerate(templates[tmplt].args) if arg_code]
+        if len(arg_codes) > 1:
+            arg_codes = list(filter(None, arg_codes))
```

Comment:
I don't fully understand the following points:

1. Why we need to pass `type_index` to `process_template`;
2. Why `process_template` is allowed to return `None`;

The need to filter out any `None` makes the code more cumbersome. If some templates are not present in the original `arg_codes` list, surely they can be removed from the `templates` list before proceeding any further?

Reply:
If we have `arg_codes ==  [[int, int]]` for example, we are going to parse arg_code[0] the number of types in each template. we can chose to return [int, int] from within `process_template` without touching it, but this will result in duplication of [int,int] the number of types in each template. or we can return None, but we will lose [int, int]. that's is why we return it only once if we didn't touch it, and for the rest we only return None.
In the following case when we have: `arg_codes == [[int, int], [T, int]]`, we can not remove the template T from templates, because it's used in arg_codes, but still what is stated before will occur for [int, int].
This will happen also if are processing template R for `arg_codes == [[int, T]]`.
I can retrieve all the `arg_code in arg_codes` that will not be processed, then add them the next time to arg_codes, and retrieve what's will not be processed and so on... but I don't think this will do any better

Reply:
So here you are essentially dealing with the situation where `arg1 = int | T`, which is a mix between a union and a template...

Reply:
Thanks @rmahjoubi! I cannot see a clear way to further improve this algorithm, unless we put some restrictions on the way the user can combine multiple `@types` headers, unions and templates.

I still think that it is useful to restructure the code a bit, in order to improve its future maintainability. For example, one could apply the original algorithm I had suggested to each of the signatures (one for each instance of @types), after adding a function that returns only the templates that are actually present in a signature!
```python
def find_templates(signature, available_templates):
    ''' Return a list of Template objects present in the given function signature. '''

def process_template(signature, template_name, d_type):
    ''' Create a new signature where the given template is substituted with a specific datatype. '''

new_signatures = []
for signature in signatures:
    tmp_signatures = (signature,)
    for t in find_templates(signature, templates):
        tmp_signatures = tuple(process_template(s, t.name, d_type) for s in tmp_signatures for d_type in t.args)
    new_signatures += tmp_signatures
```

This pseudo-code is maybe not complete, as one needs to determine what the type of `t` is supposed to be.

Reply:
Thank you, that was very helpful. I added the restrictions, now we raise an error if a template is in a uniontype. The case of multiple headers for a function works fine, since we treat each header at a time. I went back to what you suggested before.
could you check please.